### ISSUE TITLE
httpworkers massive jump

### ISSUE THREAD

**Issue by Dalacor** [e2guardian/e2guardian#515]
I have just installed E2guardian 5.31. at one of my clients on Wednesday last week. So it's been running for about 5 days (with one reboot a couple of days later ).

According to the dstats.log - the average httpworkers is less than 200 - the max in use was about 600 when I was doing some intensive testing. The max file descriptors in use was 1300 when I was doing that intensive testing.

I have configured my file descriptors soft limit to be 12 000 and my httpworkers limit to be 5000.

However today E2guardian jumped from 140 httpworkers to 5000 httpworkers forcing me to reboot the system. After rebooting the system, the httpworkers went down to 104 workers. Given that over the entire week, the maximum in use was 600 httpworkers (which was when I was testing), I don't believe that within the space of a 5 minute logging interval the entire system jumped from 140 httpworkers to 5000 which is the limit.

Has anyone seen this issue with E2g 5.x spiking httpworkers like this before? I assume that whatsapp or some website is causing e2guardian to chew httpworkers?

I am not using MITM yet as I am still testing the new system so did not want to complicate the issue with certificates and MITM etc.

**Comment by Dalacor** [comment]
I have looked through the access logs and seen that dropbox and amazonaws.com was blocked. I have unblocked amazonaws.com as this will break a lot of websites that pull content from the Amazon servers.

Don't know if dropbox or amazonaws or the two simultaneously had anything to do with the problem. Although both these websites/urls were blocked at other times.

**Comment by Anthony-76** [comment]
In /var/log/message, do you have oom error (out of memory)?
I have a problem like yours. May be your problem is oom...

Can you tell us which machine do you use (ram, cpu...) ?

**Comment by Dalacor** [comment]
Most of the computers have 4gb ram and a core i5 cpu. However I don't see that this is the problem as it does not explain why there is a massive jump from an average of 140 http workers to maxing out at 5000.

I think that something like whatsapp doesn't work well through the filters and was creating a feedback loop and chewing up the resources. I will need to check my blacklists are not causing the problem as I am using new blacklists.

**Comment by Anthony-76** [comment]
4GB of ram seems small.
How many users are filtered by E2guardian?

**Comment by Dalacor** [comment]
You serious?

According to top I am using 1.2GB and I have over 2GB free. Presumably cache has the rest. E2guardian was using 15% of that 1.2GB so basically 200mb?

Maximum httpworkers today is 500. I have never managed to get higher than 600 http workers.

The school in question would have around 20 computers max in use all day long and they would have 30 computers in the IT Suite using the Internet for a lesson and it is then it peaks to around 400-600 depending on day and what is being accessed.

I am not opposed to putting in more memory, but I cannot see that the specs are the issue as it does not explain why it would jump from 140 (which is average) right up to 5000. In the week I have been using it, it has never gone above 600 http workers.

I will see if it happens again and see what the top usage is.

**Comment by Anthony-76** [comment]
Maybe I'm wrong, but your problem seems like mine opened and closed.
Ticket number #488

**Comment by Dalacor** [comment]
I have looked at that post and I think that I should reduce my httpworkers to possibly around 1000 to 1500 as I can see that having more than this serves no purpose if I am peaking at 600 http workers and it just hogs memory.

If it happens again, I will have a look at top and see what the ram and cpu usage is. However, I don't believe that memory is my problem as it does not explain a huge jump. Your post indicates that the http workers in use did not rise, you just basically ran out of memory because your http workers was set to 10 000 which is way too high for 15 users.

My ram usage indicates that I am only using a quarter of the ram available.

**Comment by Anthony-76** [comment]
Before to reboot your system, are you able to connect via ssh or keyboard in your system,?

If yes, instead to reboot, can you type e2guardian -Q ( this command will be restart E2guardian an reinit httpworker)

Moreover, do you have information in syslog or messsage file?

**Comment by Dalacor** [comment]
Yes I can putty ssh. I normally do /usr/local/e2guardian/sbin/e2guardian -r to restart. Does this not reinit httpworkers. I am restarting e2guardian automatically at 4am in the morning to update the blacklists using the above command.

Only thing that I can see in syslog is this - e2guardian thttps connection handler caught an exception basic string::_M_construct null not valid - but that was several days before the issue. So not related. Not sure what the cause of that error is.

I have noticed that the memory usage is going up. It was 15% and is now 19%. I will take screen shots of both filtering system boxes every day and see if the memory continues to go up - this might be the cause of the problem. The first box is showing 700mb in use and I now see that is 20% of total memory, not 20% of memory in use.

I think it may be the same problem as your issue if the ram is increasing. I will check it every day and see if the same issue occurs again. I expect the box to crash on Sat/Sun then as it took 5 days last time.

**Comment by Dalacor** [comment]
I will update this weekend or Monday.

**Comment by Anthony-76** [comment]
Of my part I use MITM SSL. 
In this case, I Have lot of connections and lot of webpages are filtered by days (about 50 000 pages/day)

With this topoly E2guardian grown in memory use.

For me, this problem is appeared with 5.3.1 release. 
I have 1 or 2 times by day an out of memory  issue (oops information that I found in /var/log/messages)

To avoid this issue, I make a bash file who check every minutes (with crontab) the memory usage.
If the value is cirtical, I do via my bash file the e2guardian -Q command.

I know it's not a good workaround, but currently, I have no free time to ingestigate :(

**Comment by Dalacor** [comment]
I was just looking at my old E2guardian boxes running 3.x version and one of the computers has been up for over a year and yet ram used is just over 1GB with 3GB free. So clearly memory usage is much higher in 5.31. with 5000 httpworkers. As I don't need this many workers, I will reduce to around 1000 once I have determined that the memory is the issue.

I cannot comment on any other version as I only have 5.3.1 and 3.x version! But if you have to restart e2guardian that frequently something is wrong. It is most likely the number of workers you have and other things.

I had a look at the messages log which I did not notice. No error message at the time other than saying 5000 httpworkers, so none free. But nothing preceding this to explain jump. but I will know by this weekend as after five days I should see the same thing occur again.

**Comment by Anthony-76** [comment]
On release 4.x, I have not this issue. This appears only with relase 5 et it's seem  more recurrent with 5.3.1 release

**Comment by Dalacor** [comment]
https://github.com/e2guardian/e2guardian/pull/517 - This may be the explanation of your memory leak and possible cause of my issue. I won't update to latest version for a couple of weeks as I need to see what is causing my particular issue - memory or something else.

**Comment by Dalacor** [comment]
Checked today. Ram has gone up to 29% used but there is still well over a GB free!

HTTP Workers jumped from about 50 workers to 5000 workers from 1:20pm Friday afternoon and went back down again at 8:30pm without any reboot from me. Therefore I don't think it is the memory causing the issue. As it lasted 8 hours and most of the staff would have left by 4pm, it's obviously some kind of bug because the system would have been doing nothing for at least 4 hours.

The second box I will have to wait another couple of days to see if the system crashes on that one as we had to shut the power off earlier this week, so it's not been running long enough.

I will update the first box, but not the second and see if the issue is the version or something specific at that particular site.

**Comment by Anthony-76** [comment]
I just upgraded from release 5.3.1 to 5.3.2

I think to be able to tell  at the end of this day if the new release resolve my memory problem.

**Comment by Dalacor** [comment]
Do let me know. I am planning to upgrade the one box tonight as memory is slowly running out. But I still have no idea about the http worker jump. but the new version may have fixed this bug.

**Comment by Anthony-76** [comment]
For the moment, I have no memory issue

**Comment by Anthony-76** [comment]
Did you upgraded your release?

Of my part, I have still memory issue.

**Comment by Dalacor** [comment]
I will upgrade both boxes today and see if this fixes the httpworkers issue. We will probably have to upgrade again to fix the memory issue if latest release is not fixing the memory problem you say. Has it improved for you.

**Comment by fredbcode** [comment]
Do you have some messages in syslog?

**Comment by Dalacor** [comment]
Nothing that is relevant to the issue. I will upgrade to the latest version and reduce httpworkers to 800 as I don't seem to go above 600 (in normal usage). I will also reduce the hard and soft limits on Linux as I currently have them at 12000 and 18000. I will use the dstats log to work out the maximum in normal use and just double the count.

**Comment by Dalacor** [comment]
I will be updating E2guardian to latest version tonight to see if this fixes issue. Not had time yet. Yesterday I restarted the server at 1pm and 2 hours later it maxed out again. I am only seeing this problem at one school. The other school has not experienced this issue once. So I am now shutting down servers, security cameras and systems that are unique to the school experiencing problems to see if I can isolate the cause as its obviously a server, camera or a persons phone that is causing the problem.

Disabling transparent https proxy does not resolve the issue so we can rule out a lot of equipment that connects to the Internet using https. I will try removing transparent proxy for http as well. But the conclusion I am drawing is that the problem is not specifcally e2guardian per se as the other school is completely unaffected.

**Comment by Dalacor** [comment]
Update - I have not had a chance to update, but since switching off two servers that we don't need running 24/7 the issue seems to have disappeared so far. However, I will need to wait two weeks before concluding that the servers are responsible as the problem was very intermittent.

But despite the memory growth issue, I am satisfied that the cause of the problem is not memory related.

Has upgrading to the newest version improved the memory usage (even though not completely fixing the issue)

**Comment by Dalacor** [comment]
First site - the issue came back for an hour and went away on Thursday. But now the second site is displaying the issue - problem started at 2pm Friday afternoon and this morning 12:00 midday Saturday when I checked the http workers count is still at 5000. This is the first time that it has ever happened at the second site.

Two things - I am seeing this in the logs:

access.log - host name is showing voyager.domain.internal (which is the host name of the e2guardian proxy server) - normally I would see the host name of the client computer. For the website address I am seeing _redacted_ instead of the actual website address that the client requested.

syslog - [e2guardian 1084]: hw4964: Failed to get client's original destination ip: No such file or direct
or I get hw434 or hw1135 or hw2020 but with the exact same message.

Currently 1.4GB used, 1.4GB free and 1,1GB in buffer, CPU is pretty much idle.

I will try the latest version this weekend, otherwise I will revert back to a 4.x version as I think the problem must be with version 5.x

**Comment by fredbcode** [comment]
Are you using NTLM?

Le 30 mars 2019 13:49:21 GMT+01:00, Dalacor <notifications@github.com> a écrit :
>First site - the issue came back for an hour and went away on Thursday.
>But now the second site is displaying the issue - problem started at
>2pm Friday afternoon and this morning 12:00 midday Saturday when I
>checked the http workers count is still at 5000. This is the first time
>that it has ever happened at the second site.
>
>Two things - I am seeing this in the logs:
>
>access.log - host name is showing voyager.domain.internal (which is the
>host name of the e2guardian proxy server) - normally I would see the
>host name of the client computer. For the website address I am seeing
>http://voyager:8080 instead of the actual website address that the
>client requested.
>
>syslog - [e2guardian 1084]: hw4964: Failed to get client's original
>destination ip: No such file or direct
>or I get hw434 or hw1135 or hw2020 but with the exact same message.
>
>Currently 1.4GB used, 1.4GB free and 1,1GB in buffer, CPU is pretty
>much idle.
>
>I will try the latest version this weekend, otherwise I will revert
>back to a 4.x version as I think the problem must be with version 5.x
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/e2guardian/e2guardian/issues/515#issuecomment-478242818

-- 
Envoyé de mon appareil Android avec Courriel K-9 Mail. Veuillez excuser ma brièveté.

**Comment by Dalacor** [comment]
I am using port authentication only as well as transparent http and https proxying, No MITM or anything else complicating the issue yet.

**Comment by fredbcode** [comment]
> access.log - host name is showing voyager.domain.internal (which is the host name of the e2guardian proxy server) - normally I would see the host name of the client computer. For the website address I am seeing _redacted_ instead of the actual website address that the client requested.

This is really interesting do you have clients didn't connected transparency? And are you using Linux or FreeBSD? 
For test purpose only, can you make a try with originalip off in e2guardian.conf, I guess you are impacted with https://github.com/e2guardian/e2guardian/blob/v5.3/configs/e2guardian.conf.in#L208 (DNS performance comment)

**Comment by philipianpearce** [comment]
Is 8080 an e2g port?   Can you trigger the problem from a browser proxied to e2g by going to url _redacted_?    This could be generating a loop which would push up busy workers to max??

**Comment by Dalacor** [comment]
I have now updated to 5.3.2 last night. Will monitor ram usage increase over this week.

I have ports 8080, 8081, 8082, 8083 as the port authentication ports. port 8080 is also the transparent proxy port for http traffic obviously. I could try turning off tranparent http and see if the problem goes away as its obviously something to do with transparent filtering.

I am using Linux Slackware. Never had this issue in previous version of E2guardian (which was 3.x version).

Original destination ip is not compiled therefore its off. But I am seeing messages in the syslog regarding this. Interesting as its not compiled so should not be seeing this.  I have compiled e2guardian as follows:

./configure --prefix=/usr/local/e2guardian --enable-clamd=yes --enable-email=yes --with-proxyuser=e2guardian --with-proxygroup= e2guardian --enable-sslmitm=yes

I will do some testing this evening to see if I can trigger a loop by going to the website. I agree - there is some kind of loop happening. But not sure why.

**Comment by philipianpearce** [comment]
I have replicated a looping issue here - although I am not using an upstream squid.

See #523

**Comment by Dalacor** [comment]
I am not using squid either. Just using e2guardian.

When you say Temporary workround is to ban access from localhost, do you mean that I need to block internet access from the proxy server itself (other than forwarding traffic that is), or do you mean that I need localhost on client machines to bypass proxy server.

One thing that I have noticed is that our email system - Mdaemon - I have had to whitelist this because it works fine in Firefox but for some reason is unbelievably slow in IE and edge on W10. But for some bizarre reason is fine in IE on W7! Go figure. It may be completely unrelated, but it might explain why only my system seems to be affected as nobody else has reported this issue of massive jump in http workers. But it may just be a co-incidence because it doesn't explain why one site was affected a lot and the other site has only been affected once!

I will update you if the problem occurs again on the new version of e2g. Thank you for investigating.

**Comment by philipianpearce** [comment]
Temporary workround:  I mean adjust the firewall rules (or use iptables directly) on the server running e2g to block access (reject connections) from localhost (127.0.0.1) to the e2g proxy ports (8080:8083). 

It is also important that localhost, 127.0.0.1, (and any other services on the local LAN) are proxy bypassed on client machines so that they do not attempt to go to get these services from localhost on the e2g proxy server.

**Comment by Dalacor** [comment]
Good point. I will check the clients settings to see if proxy is bypassed for localhost etc. It may already be.  I will leave it for the moment as I have whitelisted our mail server. I want to gradually eliminate things so that I can find out what is causing it.

**Comment by Dalacor** [comment]
Either since updating to E2g 5.3.2 or since whitelisting our mail server webclient the problem has not resurfaced. However I will give it more time as it's a bit intermittent. However, memory continues to climb even with E2G 5.3.2.

I will rollout E2g 5.3.2 to other clients and see if any of them have any issues. With the problem site, I will remove the webmail site from the whitelist and see if the problem comes back to determine if the issue is with E2g 5.3.1 or that website.

**Comment by Anthony-76** [comment]
Hi,

I just realized when I execute the following command, E2guardian "eat" RAM
e2guardian -r.
I execute this command about 5-6 times by hour for dynamic configuration changing.

The reload action seems use RAM. In my case each time that I execute this command, e2guardian use about 150MB of RAM and doesn't release it after....

May be it's a normal case, maybe it's  track about a memory leak

**Comment by Dalacor** [comment]
WOW - That is exactly what I am doing. I am rebooting E2guardian every day via a script using that exact command to update the blacklists. I will turn that off for the moment and see if the memory leak goes away as result. If it does, I will log a new ticket to advise what is causing the memory leak.

My problem with the httpworkers massive jump for the time being seems to have been addressed. I believe its the email website that is the cause. But I can't test it right now as the school in question is overloaded with problems and they don't need any further hassles right now. Whitelisting the email website fixes the issue - assuming that this is the problem as I have made other changes.

**Comment by philipianpearce** [comment]
> Hi,
> 
> I just realized when I execute the following command, E2guardian "eat" RAM
> e2guardian -r.
> I execute this command about 5-6 times by hour for dynamic configuration changing.
> 
> The reload action seems use RAM. In my case each time that I execute this command, e2guardian use about 150MB of RAM and doesn't release it after....
> 
> May be it's a normal case, maybe it's track about a memory leak

@Anthony-76    Every time an -r (or -q) is issued a new master list container which contains the lists structure (not the lists themselves unless one has changed) and phrase list is created in memory.  The old list container(s) are kept until all the connections are using them are completed, then they are deleted and the memory released for other uses by e2g.    So memory should not continue to go up after every -r.  

The dstat log gives a count of the ListContainers held in memory (LCcnt) so check this.  This may go up after a -r but should return to 1 as old connections finish.   If this stays at a higher level, then this indicates that you have some clients holding connections open for an excessive period.  In this case it is not a memory leak, but normal behavior.   So, it is always best to check that there has actually been a list change before issuing a list reload (-r) and also 'save' up changes for a short period so that they are all actioned at once.

I would recommend that e2g is completely restarted (stopped and then started) once a day in order to disconnect any long-running connections and to re-set the memory usage. It is also a good opportunity to rotate the logs.

**Comment by Dalacor** [comment]
I have left the system running for over a week now with no restarts and can confirm that it is indeed the -r command that is causing the problem with memory increase. Now I see what the the ListContainer is all about. I was wondering what that was for.

What command do you recommend for completely restarting e2guardian?

**Comment by Anthony-76** [comment]
Hi,

Of my part each night after to do an update of my blacklists, I use the e2guardian -Q to restart totally E2guardian

**Comment by philipianpearce** [comment]
Sorry for the delay in repling I somehow missed this.

1. If you are starting via systemd or init.d then use the system stop command for the service.
2. Rotate log
3. Use the system start command for the service.

If you are not using systemd etc then. 
1. e2guardian -q
2. Rotate log
3. e2guardian

**Comment by philipianpearce** [comment]
Where MITM and openssl v1.1 is used see also #527 and #542 re memory leaks.

**Comment by philipianpearce** [comment]
Can I close this??

**Comment by Dalacor** [comment]
Yes. Sorry. The issue was resolved when I whitelisted our mail server website.
